File: app/src/main/java/com/octo/androidweather/CityForecastModule.java
package com.octo.androidweather;

import com.octo.presentation.CityForecastPresenterImpl;
import com.octo.presentation.DateUtilsInjector;
import com.octo.usecases.CityForecastInteractor;
import com.octo.usecases.CityForecastPresenter;
import com.octo.usecases.CityForecastRepository;

public class CityForecastModule {
    private final CityForecastController controller;

    public CityForecastModule(final MainActivity activity) {
        final CityForecastPresenter outputPort = new CityForecastPresenterImpl(activity, activity.getResources(), new DateUtilsInjector());
        final CityForecastRepository repository = new MockCityForecastRepository();
        final CityForecastInteractor interactor = new CityForecastInteractor(repository, outputPort);
        this.controller = new CityForecastController(interactor);
    }

    public CityForecastController getController() {
        return controller;
    }
}

File: app/src/main/java/com/octo/androidweather/CityForecastModule.kt
package com.octo.androidweather

import com.octo.usecases.CityForecastInteractor

import com.octo.presentation.CityForecastPresenterImpl
import com.octo.presentation.DateUtilsWrapper
import com.octo.repository.network.NetworkModule
import com.octo.repository.network.WeatherNetworkRepository

import com.octo.usecases.CityForecastPresenter
import okhttp3.HttpUrl


class CityForecastModule(activity: MainActivity) {
    val controller: CityForecastController

    init {
        val outputPort: CityForecastPresenter =
            CityForecastPresenterImpl(activity, DateUtilsWrapper())
        val interactor = CityForecastInteractor(
            WeatherNetworkRepository(
                NetworkModule().getRetrofit(
                    HttpUrl.get(BuildConfig.WEATHER_BASE_URL)
                )
            ), outputPort
        )
        controller = CityForecastController(interactor)
    }
}